Assets/Tile walking mini game/Scripts/TileWalkingUI.cs

```diff
@@ -1,47 +1,85 @@
 using System.Collections;
 using System.Collections.Generic;
 using UnityEngine;
+using UnityEngine.SocialPlatforms.Impl;
 using UnityEngine.UI;
 
 public class TileWalkingUI : MonoBehaviour
 {
+    [Header("Dependencies")]
+
     [SerializeField] private GridManager gridManager;
 
+    [SerializeField] private Timer timer;
+
     [SerializeField] private PlayerControls playerControls;
 
-    [SerializeField] private Text gameOverText;
+    [Header("UI")]
 
-    [SerializeField] private Text winText;
+    [SerializeField] private Text endGameText;
 
     [SerializeField] private Button gameOverButton;
 
+    [SerializeField] private Text currentScoreText;
+
+    [SerializeField] private Text timerText;
+
+
     private void Awake()
     {
         gameOverButton.onClick.RemoveAllListeners();
         gameOverButton.onClick.AddListener(gridManager.ResetGrid);
         gameOverButton.onClick.AddListener(playerControls.ResetPlayerMovement);
         gameOverButton.onClick.AddListener(DisableAllUIElements);
 
-        gridManager.WinEvent += ToggleWinUI;
-        gridManager.LoseEvent += ToggleGameOverUI;
+        gridManager.GameEventUpdatedEvent += UpdateUI;
+        gridManager.TileStatusChangedEvent += UpdateCurrentScore;
+
+        timer.TimerUpdatedEvent += UpdateTimer;
     }
 
-    public void ToggleGameOverUI()
+    private void UpdateUI(GameState gameState)
     {
-        gameOverText.gameObject.SetActive(!gameOverText.gameObject.activeSelf);
-        gameOverButton.gameObject.SetActive(!gameOverButton.gameObject.activeSelf);
+        if (gameState.CheckCurrentState("full win"))
+            ToggleEndGameUI(message: "You won!");
+
+        else if (gameState.CheckCurrentState("partial win"))
+            ToggleEndGameUI(message: "You (kinda) won.");
+
+        else if (gameState.CheckCurrentState("lose"))
+            ToggleEndGameUI(message: "You lost.");
+    }
+
+    private void ToggleEndGameUI(string message)
+    {
+        endGameText.text = message;
+        endGameText.gameObject.SetActive(true);
+        gameOverButton.gameObject.SetActive(true);
+    }
+
+    private void UpdateCurrentScore(Tile tile)
+    {
+        if(tile.TileStatus != TileStatus.Touched)
+        {
+            return;
+        }
+
+        int number = 5 + gridManager.touchedTiles_Percent;
+
+        if(number > 100)
+            number = 100;
+
+        currentScoreText.text = number.ToString() + "%";

```

Comment:
You don't need to call `ToString` here, as concatenating the `"%"` with it automatically converts it to a string. That's why you see a lot of oldschool programmers do this: `number + ""` when they want to convert to string rather than using the c# method. I like the c# method but there's no point here as you're concatenating already anyway.

Reply:
Indeed.